### DISCUSSION TITLE
ZFS LOAD-KEY ERROR

### DISCUSSION THREAD

**Discussion by vulg123** [openzfs/zfs#16349] [Q&A]
Good morning
Please guys, after enter this `sudo zfs load-key rpool/ROOT/ubuntu_uy913x0` 
I got this output `Key load error: Keys must be loaded for encryption root of 'rpool/ROOT/ubuntu_uy913x' (rpool)`.

Please I dont know how to solve it

Please
Please

**Comment by gmelikov** [comment]
https://github.com/openzfs/zfs/blob/f7d8b1333699957de59d72fc57c9a771951f6582/lib/libzfs/libzfs_crypto.c#L1323-L1324

you need to load-key on `rpool` itself, not child dataset.

Offtopic, but it's not the best practice to use root dataset (rpool itself) for anything, next time it's better to encrypt at least at `rpool/ROOT` level.

**Reply by vulg123** [reply]
Hello I tried, and they gave me this error

 'Key load error: Failed to open key material  file: No such file or directory' 
That is what I get

**Reply by IvanVolosyuk** [reply]
It looks like your key location is a file and the file doesn't exist on this system. You should try to find the key and may be change the key location property.

**Reply by vulg123** [reply]
My question is how to do ? How to find the file?
How to find a key ?

I've tried everything I know, and isn't working

**Reply by IvanVolosyuk** [reply]
What does the ```zfs list -o name,encryption,encryptionroot,keylocation,keyformat -r rpool``` say?

**Comment by meonkeys** [comment]
I'm getting [the same error](https://askubuntu.com/questions/1547972/22-04-lts-full-disk-encrypted-zfs-system-boot-drops-to-initramfs-after-linux-ker) as @vulg123 .

How might I change the way the 6.x kernel boots? It feels like it has to be some DKMS / initial ramdisk config / script that tells it how to make sure `keystore` is properly mounted, then prompts me for the password.

@vulg123 , yours is an Ubuntu system, yes? How did you install Ubuntu originally? Did you do any upgrades (say, from 22.04 to 24.04?) Did you originally have a 5.x kernel and are now trying to use a 6.x kernel? Do you have ZFS 2.3.0?

```
$ zfs list -o name,encryption,encryptionroot,keylocation,keyformat -r rpool
NAME                                              ENCRYPTION   ENCROOT  KEYLOCATION                            KEYFORMAT
rpool                                             aes-256-gcm  rpool    file:///run/keystore/rpool/system.key  raw
rpool/ROOT                                        aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw                          aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/srv                      aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/usr                      aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/usr/local                aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var                      aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/games                aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/lib                  aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/lib/AccountsService  aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/lib/NetworkManager   aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/lib/apt              aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/lib/dpkg             aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/log                  aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/mail                 aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/snap                 aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/spool                aes-256-gcm  rpool    none                                   raw
rpool/ROOT/ubuntu_2bfkjw/var/www                  aes-256-gcm  rpool    none                                   raw
rpool/USERDATA                                    aes-256-gcm  rpool    none                                   raw
rpool/USERDATA/root_mxx4do                        aes-256-gcm  rpool    none                                   raw
rpool/keystore                                    off          -        none                                   none
rpool/var                                         aes-256-gcm  rpool    none                                   raw
rpool/var/lib                                     aes-256-gcm  rpool    none                                   raw
rpool/var/lib/docker                              aes-256-gcm  rpool    none                                   raw
```

It looks like something changed as far as key handling since I originally installed Ubuntu on this laptop, and now I'm trying to load a kernel that I think System76 created but supports ZFS differently? And I think ZFS itself upgraded along the way. Looks like I have ZFS 2.3.0 now.
